sam_view.c

```diff
@@ -68,13 +73,58 @@ typedef struct samview_settings {
     hts_filter_t *filter;
     int remove_flag;
     int add_flag;
+    auxhash_t remove_tag;
+    auxhash_t keep_tag;
 } samview_settings_t;
 
 
 // TODO Add declarations of these to a viable htslib or samtools header
 extern const char *bam_get_library(sam_hdr_t *header, const bam1_t *b);
 extern int bam_remove_B(bam1_t *b);
 
+// Copied from htslib/sam.c.
+// TODO: we need a proper interface to find the length of an aux tag,
+// or at the very make exportable versions of these in htslib.
+static inline int aux_type2size(uint8_t type)
+{
+    switch (type) {
+    case 'A': case 'c': case 'C':
+        return 1;
+    case 's': case 'S':
+        return 2;
+    case 'i': case 'I': case 'f':
+        return 4;
+    case 'd':
+        return 8;
+    case 'Z': case 'H': case 'B':
+        return type;
+    default:
+        return 0;
+    }
+}
+
+// Copied from htslib/sam.c.
+static inline uint8_t *skip_aux(uint8_t *s)
```

Comment:
The [version of this in HTSlib](https://github.com/samtools/htslib/blob/4811c2d8/sam.c#L4093) is careful not to fall off the end of the BAM record.  Was this from an earlier, less safe version?

The comment is correct, an HTSlib interface to iterate through the aux tags would be useful.  Or for a future update we could consider adding something to do the aux tag editing part of this PR as a handy higher-level function.

Reply:
See above - it's just that Htslib has moved on in the last 6 years since I copied these functions over.

I could create an htslib PR to make them public, but I don't particularly wish to slow up merging even more than it's already taken.